### PR TITLE
normalize handling of $! inside cfn tags like !Sub + tests

### PR DESCRIPTION
**Pull request by tavisrudd**
I also added some safe-guards around string type assumptions in calls to `visitString`;

### CODE REVIEW

**Comment by tavisrudd** [comment]
@Kazzer this will give you better error handling in the `!Sub` thing you hit the other day.

File: src/preprocess/visitor.ts [hunk 1]
```diff
@@ -475,13 +501,26 @@ export class Visitor {
...
+      if (!_.isString(subArg)) {
+        throw new Error(
+          `Invalid argument to !Sub: expected string, got ${node.data[0]}`);
+      }
+      return new yaml.Sub(this.visitSubStringTemplate(subArg, path, env));
     } else if (_.isArray(node.data) && node.data.length === 2) {
+      const subArg0 = this.visitNode(node.data[0], path, env);
+      if (!_.isString(subArg0)) {
+        throw new Error(
+          `Invalid argument to !Sub: expected string, got ${node.data[0]}: ${subArg0}`);
```
Comment (jpb) [inline src/preprocess/visitor.ts]:
```suggestion
          `Invalid argument to !Sub: expected first member of list to be string, got ${node.data[0]}: ${subArg0}`);
```

File: src/preprocess/visitor.ts [hunk 2]
```diff
@@ -186,7 +186,15 @@ export class Visitor {
     } else if (node instanceof yaml.Sub) {
       return this.visitSub(node, path, env);
     } else {
-      return node.update(this.visitNode(node.data, path, env));
+      if (_.isArray(node.data)
+        && node.data.length === 1
+        && node.data[0] instanceof yaml.Tag) {
+        // unwrap Tags like !ImportValue [!$ someVar]
+        return node.update(this.visitNode(node.data[0], path, env));
```
[the remark this answers is not in the document]
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: Yes.

File: src/preprocess/visitor.ts [hunk 3]
```diff
@@ -446,16 +458,30 @@ export class Visitor {
...
+
+    return new yaml.Ref(this.maybeRewriteRef(ref, path, env));
   }
 
   visitGetAtt(node: yaml.GetAtt, path: string, env: Env): yaml.GetAtt {
-    if (_.isArray(node.data)) {
-      const argsArray = _.clone(node.data);
+    const arg = this.visitNode(node.data, path, env);
+    if (!(_.isString(arg)
+      || (_.isArray(arg) && _.isString(arg[0])))) {
```
[the remark this answers is not in the document]
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: I suppose so from CFN's viewpoint. iidy won't blow up though, which is why I left it off.

[the remark this answers is not in the document]
Comment (tavisrudd) [reply src/preprocess/visitor.ts]: I think we should leave it as and let CFN warn about the rest.

**Pull request merged**